### PR TITLE
Reimplement dataset

### PR DESCRIPTION
**Pull request by YinengWang**
1. Implemented FastDataLoader which integrates a noise generator
2. Added a test case for the noise generator

### CODE REVIEW

**Comment by Alfred-N** [comment]
Great work!! Just one question: does it give the right batch inds to input into criterion() when shuffle=True?

`inds = np.arange(batch_idx * config.batch_size, min((batch_idx+1) * config.batch_size, train_loader.dataset_len))`

This is a smart solution! However I'm afraid that when the dataloader shuffles every epoch (or batch?) that these won't be the right indices anymore. Is there any way to extract the order in which the dataloader is shuffeling?

**Comment by YinengWang** [comment]
> Is there any way to extract the order in which the dataloader is shuffling?

Thank you for pointing it out! It does not reflect the original indices if the dataloader shuffles the input each epoch. Changes should be made. Working on it.

**Comment by YinengWang** [comment]
> > Is there any way to extract the order in which the dataloader is shuffling?
> 
> Thank you for pointing it out! It does not reflect the original indices if the dataloader shuffles the input each epoch. Changes should be made. Working on it.

Done

**Review by Alfred-N — COMMENTED** [review]
Yay, well done!

**Comment by rlindsberg** [comment]
Testing oon server2 now

**Review by rlindsberg — CHANGES_REQUESTED** [review]
@YinengWang 
Can you look at this?
Skærmbillede 2021-05-14 kl  17 04 59

**Comment by YinengWang** [comment]
It's caused by trying to save the graph when the folder doesn't exist. Just fixed it.

**Review by rlindsberg — APPROVED** [review]
Test passed. Well done!

Skærmbillede 2021-05-14 kl  17 14 32

myplot1

myplot2
